### PR TITLE
Test extensions function as intended when installed in subdomains only

### LINKED ISSUE #890: Enabling extensions in specific domains after they are installed in root
While root permissions are required to install an extension, in theory, permissions could only be awarded to an extension in a subdomain. We must make sure our extensions work in that scenario, and while I believe they should, as far as I can see we don't have tests along those lines.

### PR DESCRIPTION
**Pull request by area**
Closes #890

### CODE REVIEW

File: contracts/extensions/FundingQueue.sol [hunk 1]
```diff
@@ -283,25 +283,45 @@ contract FundingQueue is ColonyExtension, PatriciaTreeProofs {
...
 
-    if (proposal.totalPaid == proposal.totalRequested) {
-      proposal.state = ProposalState.Completed;
-
-      queue[HEAD] = queue[_id];
-      delete queue[_id];
-
-      //  May be the null proposal, but that's ok
-      proposals[queue[HEAD]].lastUpdated = updateTime;
+    try
```
Comment (kronosapiens) [inline contracts/extensions/FundingQueue.sol:286]: As per our discussion, we discussed not using `try/catch` and instead checking the permission explicitly:

```
if (!colony.hasUserRole(address(this), proposal.domainId, FUNDING_ROLE)) {
  // Cancel proposal
}
```

I think that'll be more succinct and readable.

Comment (area) [reply contracts/extensions/FundingQueue.sol:286]: Hm, I was writing a big long answer saying how that's how I started, but doing this check requires six external calls, but while writing it I've realised you are right, this is enough. That domain does indeed have to be the domain we have an explicit permission in, and so we don't have to check the inheritance etc, which can be left to be done in the main call.

Comment (kronosapiens) [reply contracts/extensions/FundingQueue.sol:286]: Fab u lous

File: test/extensions/funding-queue.js [hunk 2]
```diff
@@ -793,4 +810,91 @@ contract("Funding Queues", (accounts) => {
...
+
+      // Proposal is cancelled
+      const proposal = await fundingQueue.getProposal(proposalId);
+      expect(proposal.state).to.eq.BN(STATE_CANCELLED);
+
+      // No tokens transferred
+      const balanceAfter = await colony.getFundingPotBalance(2, token.address);
+      const amountTransferred = balanceBefore.sub(balanceAfter);
+      const expectedTransferred = new BN("0");
+      expect(amountTransferred).to.eq.BN(expectedTransferred);
```
Comment (kronosapiens) [inline test/extensions/funding-queue.js:897]: Just do `to.be.zero;`

File: contracts/extensions/FundingQueue.sol [hunk 3]
```diff
@@ -282,6 +283,15 @@ contract FundingQueue is ColonyExtension, PatriciaTreeProofs {
     proposal.lastUpdated = updateTime;
 
     assert(proposal.totalPaid <= proposal.totalRequested);
+    address colonyAuthority = colony.authority();
+
+    // Check if the extension has the permissions to do this
+    // If not, cancel the proposal so others aren't blocked
+    if (!colony.hasUserRole(address(this), proposal.domainId, ColonyRole.Funding)) {
```
[the remark this answers is not in the document]
Comment (area) [reply contracts/extensions/FundingQueue.sol:290]: No, that's absolutely the best way. I'd simply forgotten it was possible :sweat:

File: contracts/extensions/FundingQueue.sol [hunk 4]
```diff
@@ -41,6 +41,7 @@ contract FundingQueue is ColonyExtension, PatriciaTreeProofs {
   uint256 constant UINT256_MAX = (2 ** 256) - 1;
   uint256 constant STAKE_FRACTION = WAD / 1000; // 0.1%
   uint256 constant COOLDOWN_PERIOD = 14 days;
+  uint8 constant FUNDING_ROLE = 5;
```
[the remark this answers is not in the document]
Comment (area) [reply contracts/extensions/FundingQueue.sol:44]: Nothing at all, a leftover from when I was trying to do this without importing `ColonyDataTypes`.

File: contracts/extensions/FundingQueue.sol [hunk 5]
```diff
@@ -282,6 +283,15 @@ contract FundingQueue is ColonyExtension, PatriciaTreeProofs {
     proposal.lastUpdated = updateTime;
 
     assert(proposal.totalPaid <= proposal.totalRequested);
+    address colonyAuthority = colony.authority();
```
[the remark this answers is not in the document]
Comment (area) [reply contracts/extensions/FundingQueue.sol:286]: Another leftover from trying to avoid using `ColonyDataTypes` :smile:

**Review by kronosapiens — APPROVED** [review]

**Pull request merged**